src/components/OpenSeadragonComponent.js

```diff
@@ -100,13 +107,38 @@ function OpenSeadragonComponent({
       viewport.setFlip(viewerConfig.flip);
     }
 
-    if (viewerConfig.bounds && !viewerConfig.x && !viewerConfig.y && !viewerConfig.zoom) {
-      const rect = new Openseadragon.Rect(...viewerConfig.bounds);
-      if (rect.equals(viewport.getBounds())) {
+    const bounds = viewerConfig.bounds || worldBounds;
+    if (bounds && !viewerConfig.x && !viewerConfig.y && !viewerConfig.zoom) {
+      const rect = new Openseadragon.Rect(...bounds);
+      if (!rect.equals(viewport.getBounds())) {
         viewport.fitBounds(rect, false);
       }
     }
-  }, [initialViewportSet, setInitialBounds, viewerConfig, viewerRef]);
+  }, [initialViewportSet, setInitialBounds, viewerConfig, viewerRef, worldBounds]);
+
+  useEffect(() => {
+    if (!osdConfig.preserveViewport) return;
+    if (!viewerConfig?.worldBounds || !worldBounds) return;
+
+    const viewer = viewerRef.current;
+    if (!viewer) return;
+    const { viewport } = viewer;
+
+    const [_x, _y, width, height] = viewerConfig.worldBounds;
+    const [_x1, _y1, width1, height1] = worldBounds;
+
+    const previousAspectRatio = (1.0 * width) / height;
+    const newAspectRatio = (1.0 * width1) / height1;
+
+    if (previousAspectRatio < 0.75 * newAspectRatio || previousAspectRatio > 1.25 * newAspectRatio) {
+      console.log("aspect ratio changed, updating viewport");
+      const rect = new Openseadragon.Rect(...worldBounds);
+      if (!rect.equals(viewport.getBounds())) {
+        viewport.fitBounds(rect, false);
+      }
+    }
+
+  }, [osdConfig, viewerConfig, worldBounds, viewerRef]);
```

Comment:
if `worldBounds` is an array, I think this effect is always going to run because the object reference will change. I think the approach is sound, but we'd need to unroll the array into integers, right?